### PR TITLE
feat: QdrantIndex

### PR DESCRIPTION
**Pull request by Anush008**
## Description
This PR adds support for [Qdrant](http://qdrant.tech/) to be used as a supported index in Semantic Router.

## Testing
The integration tests have been configured to run using an in-memory Qdrant instance if `qdrant-client` is installed. Skipped otherwise.

### CODE REVIEW

File: semantic_router/index/qdrant.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    )
+    path: Optional[str] = Field(
+        default=None,
+        description="Persistence path for Qdrant local",
+    )
+    grpc_options: Optional[Dict[str, Any]] = Field(
+        default=None,
+        description="Options to be passed to the low-level Qdrant GRPC client, if used.",
+    )
+    size: Union[int, None] = Field(
```
Comment (jamescalam) [inline semantic_router/index/qdrant.py:70]: to keep attributes aligned between different indexes could we rename to `dimensions`?

File: semantic_router/index/qdrant.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        default=None,
+        description="Embedding dimensions. Defaults to the embedding length of the configured encoder.",
+    )
+    distance: str = Field(
```
Comment (jamescalam) [inline semantic_router/index/qdrant.py:74]: should be `metric`?

File: semantic_router/index/qdrant.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+DEFAULT_UPLOAD_BATCH_SIZE = 100
+SCROLL_SIZE = 1000
+SR_UTTERANCE_PAYLOAD_KEY = "sr_utterance"
+SR_ROUTE_PAYLOAD_KEY = "sr_route"
+
+
+class QdrantIndex(BaseIndex):
+    "The name of the collection to use"
+
+    collection_name: str = Field(
```
Comment (jamescalam) [inline semantic_router/index/qdrant.py:18]: could we rename to `index_name` for alignment between indexes?

File: semantic_router/index/qdrant.py [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+        description="Options to be passed to the low-level Qdrant GRPC client, if used.",
+    )
+    size: Union[int, None] = Field(
+        default=None,
+        description="Embedding dimensions. Defaults to the embedding length of the configured encoder.",
+    )
+    distance: str = Field(
+        default="Cosine", description="Distance metric to use for similarity search."
+    )
+    collection_options: Optional[Dict[str, Any]] = Field(
```
Comment (jamescalam) [inline semantic_router/index/qdrant.py:77]: given we are renaming `collection_name` to `index_name`, maybe we can rename this? Maybe `config`? We can apply the same to the `PineconeIndex` where we currently have `ServerlessSpec`

File: semantic_router/index/qdrant.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+                "You can install it with: "
+                "`pip install 'semantic-router[qdrant]'`"
+            ) from e
+
+    def _init_collection(self) -> None:
+        from qdrant_client import QdrantClient, models
+
+        self.client: QdrantClient
+        if not self.client.collection_exists(self.collection_name):
+            if not self.dimensions:
```
Comment (jamescalam) [inline semantic_router/index/qdrant.py:119]: I think there's a typo here? You set `size` as the param name for the class but here it is `dimensions` — after changing `size` to `dimensions` this should be okay

File: semantic_router/index/qdrant.py [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+                raise ValueError(
+                    "Cannot create a collection without specifying the dimensions."
+                )
+
+            self.client.create_collection(
+                collection_name=self.collection_name,
```
Comment (jamescalam) [inline semantic_router/index/qdrant.py:125]: will be renamed to `self.index_name`

File: semantic_router/index/qdrant.py [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+                vectors_config=models.VectorParams(
+                    size=self.dimensions,
+                    distance=self.distance,  # type: ignore
```
Comment (jamescalam) [inline semantic_router/index/qdrant.py:128]: will be renamed to `self.metric`

File: semantic_router/index/qdrant.py [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+                ),
```
Comment (jamescalam) [inline semantic_router/index/qdrant.py:129]: Given that qdrant has slightly different metrics to the other indexes, I think we need a `Metric` type in `schema.py` which will include standard string values that will be used by the library and converted into the correct metric string for each service. Something like:

```python
class Metric(Enum):
    COSINE = "cosine"
    DOTPRODUCT = "dotproduct"
    EUCLIDEAN = "euclidean"
    MANHATTEN = "manhattan"
```

File: semantic_router/index/qdrant.py [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+            )
+
+    def add(
+        self,
+        embeddings: List[List[float]],
+        routes: List[str],
+        utterances: List[str],
+        batch_size: int = DEFAULT_UPLOAD_BATCH_SIZE,
+    ):
+        self.dimensions = self.dimensions or len(embeddings[0])
```
Comment (jamescalam) [inline semantic_router/index/qdrant.py:140]: noting that this isn't aligned with `self.size` — should be handled with switch from `self.size` to `self.dimensions`

File: semantic_router/index/qdrant.py [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+        """
+
+        import grpc
+
+        results = []
+        next_offset = None
+        stop_scrolling = False
+        while not stop_scrolling:
+            records, next_offset = self.client.scroll(
+                self.collection_name,
```
Comment (jamescalam) [inline semantic_router/index/qdrant.py:171]: should be changed to `self.index_name`

File: semantic_router/index/qdrant.py [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+            (x.payload[SR_ROUTE_PAYLOAD_KEY], x.payload[SR_UTTERANCE_PAYLOAD_KEY])
+            for x in results
+        ]
+        return route_tuples
+
+    def delete(self, route_name: str):
+        from qdrant_client import models
+
+        self.client.delete(
+            self.collection_name,
```
Comment (jamescalam) [inline semantic_router/index/qdrant.py:194]: should be `self.index_name`

File: semantic_router/index/qdrant.py [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+                    models.FieldCondition(
+                        key=SR_ROUTE_PAYLOAD_KEY,
+                        match=models.MatchText(text=route_name),
+                    )
+                ]
+            ),
+        )
+
+    def describe(self) -> dict:
+        collection_info = self.client.get_collection(self.collection_name)
```
Comment (jamescalam) [inline semantic_router/index/qdrant.py:206]: `self.collection_name` -> `self.index_name`

File: semantic_router/index/qdrant.py [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+    def query(self, vector: np.ndarray, top_k: int = 5) -> Tuple[np.ndarray, List[str]]:
+        results = self.client.search(
+            self.collection_name, query_vector=vector, limit=top_k, with_payload=True
+        )
+        scores = [result.score for result in results]
+        route_names = [result.payload["sr_route"] for result in results]
+        return np.array(scores), route_names
+
+    def delete_index(self):
+        self.client.delete_collection(self.collection_name)
```
Comment (jamescalam) [inline semantic_router/index/qdrant.py:223]: `self.collection_name` -> `self.index_name`

File: semantic_router/index/qdrant.py [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+
+    def __len__(self):
+        return self.client.get_collection(self.collection_name).points_count
```
Comment (jamescalam) [inline semantic_router/index/qdrant.py:226]: `self.collection_name` -> `self.index_name`

**Review by jamescalam — CHANGES_REQUESTED** [review]
hi @Anush008 thanks for the PR this is great! I have a few comments, could we fix those and then we should be good to run tests and merge all being okay, thanks!

File: semantic_router/index/qdrant.py (hunk 2 above)
Comment (Anush008) [reply semantic_router/index/qdrant.py]: I was wondering if we should go with index provider-specific lingo or the framework's. Will go with SR's.

File: semantic_router/index/qdrant.py (hunk 1 above)
Comment (Anush008) [reply semantic_router/index/qdrant.py]: I was wondering if we should go with index provider-specific lingo or the framework's. Will go with SR's.

File: semantic_router/index/qdrant.py (hunk 8 above)
Comment (Anush008) [reply semantic_router/index/qdrant.py:130]: https://github.com/aurelio-labs/semantic-router/pull/207/commits/a9bcc0ee2c56f02cca82e94f0dd9cdb498db081f

Comment (Anush008) [reply semantic_router/index/qdrant.py:130]: https://github.com/aurelio-labs/semantic-router/pull/207/commits/78a2daec28663d2025d7d6cf40e44300786f58b6

File: semantic_router/index/qdrant.py (hunk 4 above)
Comment (Anush008) [reply semantic_router/index/qdrant.py:79]: `config` would be too generic, don't you think? Since these are options for collection creation specifically.

Comment (jamescalam) [reply semantic_router/index/qdrant.py:79]: Possibly, what would you recommend? Is it within `collection_options` that I would specify `hnsw_config`, `wal_config`, `optimizers_config`, `shard_number`, `on_disk_payload`, and `quantization_config`?

We want something that can align with other vec dbs or storage options, my thought is that `collection_options` is specific to qdrant which is okay but for people switching between `PineconeIndex` and `QdrantIndex` who are not familiar with both the changing parameter names could be confusing

Comment (Anush008) [reply semantic_router/index/qdrant.py:79]: > Is it within collection_options that I would specify hnsw_config, wal_config, optimizers_config, shard_number, on_disk_payload, and quantization_config

Yes.

I can rename it to `config` if that would be beneficial as you mentioned.

Comment (Anush008) [reply semantic_router/index/qdrant.py]: https://github.com/aurelio-labs/semantic-router/pull/207/commits/622d422c61de50ac710772fce1c89eb1fea61e7e

**Review by mesax1 — APPROVED** [review]
LGTM.
Passes tests, and the naming conventions follow SemanticRouter conventions of index, metadata, query, dimensions, and config.

**Review by jamescalam — APPROVED** [review]
nice, thanks all, we appreciate the PR @Anush008!

**Pull request merged**